Updater/Server.cs

```diff
@@ -256,76 +269,113 @@ private static void MetadataHandler(DataPacket dataPacket, ICommunicator communi
-            {
-                Directory.CreateDirectory(Path.GetDirectoryName(tempFilePath)!);
-                File.WriteAllText(tempFilePath, serializedDifferences);
-                Server.UpdateUILogs($"Differences file saved to {tempFilePath}");
-                Trace.WriteLine($"[Updater] Differences file saved to {tempFilePath}");
-            }
-            catch (Exception ex)
-            {
-                Trace.WriteLine($"[Updater] Error saving differences file: {ex.Message}");
-            }
+                FileContent fileContentToSend = new FileContent(
+                        "invalidFileNames.list",
+                        Utils.SerializeObject(invalidFileNames)
+                        );
 
-            // Prepare data to send to client
-            List<FileContent> fileContentsToSend = new List<FileContent>
+                DataPacket dataPacketToSend = new DataPacket(
+                        DataPacket.PacketType.InvalidSync,
+                        new List<FileContent> { fileContentToSend }
+                        );
+
+                string serializedDataPacket = Utils.SerializeObject(dataPacketToSend);
+
+                try
                 {
-                    // Added difference file to be sent to client
-                    new FileContent("differences.xml", serializedDifferences)
-                };
+                    Server.UpdateUILogs($"Sending files to client and waiting to recieve files from client {clientId}");
+                    communicator.Send(serializedDataPacket, "Server", clientId);
 
-            // Retrieve and add unique server files to fileContentsToSend
-            foreach (string filename in comparerInstance.UniqueServerFiles)
+                    // End the sync up
+                    server.CompleteSync();
+                }
+                catch (Exception ex)
+                {
+                    Trace.WriteLine($"[Updater] Error sending data to client: {ex.Message}");
+                }
+            }
+
+            // If sync up is valid, send differences file to client
+            else
             {
-                string filePath = Path.Combine(Server.s_serverDirectory, filename);
-                string? content = Utils.ReadBinaryFile(filePath);
+                // Serialize and save differences to C:\temp\ folder
+                string serializedDifferences = Utils.SerializeObject(differences);
+                string tempFilePath = @$"{Server.s_serverDirectory}\differences.xml";
 
-                if (content == null)
+                if (string.IsNullOrEmpty(serializedDifferences))
                 {
-                    Trace.WriteLine($"Warning: Content of file {filename} is null, skipping.");
-                    continue; // Skip to the next file instead of throwing an exception
+                    Trace.WriteLine("[Updater] Serialization of differences failed or resulted in an empty string.");
+                    return; // Exit if serialization fails
```

Comment:
General doubt: Will the client release semaphore due to returning? Or need to complete sync up here as well?

Reply:
Also, is the semaphore released for next client if in between the sync up of previous client some error happens?

Reply:
Serialized string cannot be empty
To get more understanding, you can read and run this repo: 
[](https://github.com/Amithabh-A/SerializerTester)

Reply:
If exceptions are happening, I would encourage you to write a function `ExceptionHandler` which call `CompleteSync()` along with writing exception messages in trace